File: cpp/bj/9663_nqueen.cpp
#include <iostream>

int queens[15];
int columns[15];
int d1[27];
int d2[27];

int solve(int N, int row)
{
  if (N == row) return 1;

  int ret = 0;
  for (int col = 0; col < N; col++)
  {
    // check
    if (columns[col] || d1[row + col] || d2[row + (N - col)]) continue;
    
    columns[col] = 1;
    d1[row + col] = 1;
    d2[row + (N - col)] = 1;
    ret += solve(N, row + 1);
    columns[col] = 0;
    d1[row + col] = 0;
    d2[row + (N - col)] = 0;
  }
  return ret;
}

int main(void)
{
  int N;
  std::cin >> N;

  int answer;
  answer = solve(N, 0);
  std::cout << answer;
}

File: ALGO/6BACK-TRACKING/9663_N-Queen.py
from sys import stdin, stdout

input = stdin.readline


def println(s):
    stdout.write(f"{s}\n")


def possible(row, col):
    global N, queens, used
    if used["column"][col]:
        return False
    if used["right_diagonal"][(N - col) + row]:
        return False
    if used["left_diagonal"][col + row]:
        return False
    return True


def update(row, col, put):
    global used
    used["column"][col] = put
    used["right_diagonal"][(N - col) + row] = put
    used["left_diagonal"][col + row] = put


def recur(row):
    global N, queens
    if row == N:
        return 1
    ret = 0
    for col in range(N):
        if possible(row, col):
            queens[row] = col
            update(row, col, True)
            ret += recur(row + 1)
            update(row, col, False)
    return ret


N = int(input())
queens = [-1] * N
used = {
    "column": [False] * N,
    "right_diagonal": [False] * (2 * N + 1),
    "left_diagonal": [False] * (2 * N + 1),
}
print(recur(0))
